### ISSUE TITLE
Eel Mania (again)

### ISSUE THREAD

**Issue by mxnmnm** [itsmeow/betteranimalsplus#167]
**Describe the bug**
Eelmania, I'm sure you remember

**To Reproduce**
1. exist near a body of fresh water (specifically rivers)

**Expected behavior**
Eel population being under control

**Screenshots**
![2020-04-19_19 00 29](https://user-images.githubusercontent.com/8508976/79707081-2a8c3080-8270-11ea-95ae-d53aa8e456c9.png)

**Versions:**
 - OS: Windows 10
 - MC Version: 1.12.2
 - Mod Version: 9.0.1

**Additional context**
gl on trying to fix it (again)

**Comment by itsmeow** [comment]
FUCK EELS

**Comment by itsmeow** [comment]
LITERALLY HOW

**Comment by mxnmnm** [comment]
Yea no clue lol

**Comment by mxnmnm** [comment]
oh I have an idea, try it with the blood moon mod during a blood moon since that's when I noticed this happen

**Comment by mxnmnm** [comment]
Let me test real quick lol

**Comment by mxnmnm** [comment]
Nope

**Comment by mxnmnm** [comment]
F

**Comment by SandwichHorror** [comment]
Can you post the mod list, @mxnmnm ?

**Comment by mxnmnm** [comment]
https://pastebin.com/TsfLJ24f

**Comment by MechanosG** [comment]
Heya.  I just updated Better Animals Plus to resume work on a mod pack..  noticed random lag.  Then I noticed entity fish absolutely everywhere in a nearby pond lol.  Ran lag goggles, and it says they're from this mod.  Mostly Freshwater Eels.  I'd like to help provide information, if I can.

I looked through the mod list @mxnmnm posted, and one mod that sticks out that we both have, is "In Control".  That mod customizes vanilla mob spawning via json defined rules.  Could it be the cause of this eel spam?  For example, I have a rule that forbids passive mob spawning once 15 mobs per player exist.  I'm wondering if attempting to reject a passive spawn, somehow makes more spawn, in this extremely specific case?  I know Mowzie's Mobs had a similar problem with the flying lantern mob spawning in huge numbers, and I wondered the same thing.  I just don't know if they ever figured out what was the cause.

Anyway, lemme know if I can help in any way.  I'm thinking about temporarily removing the rule and/or temporarily removing In Control then seeing if the eel spam stops.

**Comment by mxnmnm** [comment]
If incontrol is the issue then you could easily use it to fix the issue without removing it. If it is the issue some sort of warning would probably be useful to people wanting to use both so they can fix the issue themselves.

**Comment by MechanosG** [comment]
I'm not so certain.  As I already said, I had an existing rule to deny passive spawns once there was 15 of any passive mob (and I've confirmed that it works on other mobs), but freshwater eels rapidly spawn regardless.

Btw, removing In Control fixed the issue.  Unfortunately In Control plays a fairly important part in my modpack, so I'm going to have to test your theory to see if I can find some way to make a rule that will work for the eels.  Perhaps I'll have better luck if I deny the specific mob instead of using the passive tag.  I ended up having to do that with squids now that I think about it.  If it works, maybe I'll share the rule if anyone's interested.

**Comment by mxnmnm** [comment]
Try this (it goes in spawn.json)
[{
	"mob": "betteranimalsplus:eel_freshwater",
	"maxcount": {
	"amount": 3,
	"mob": "betteranimalsplus:eel_freshwater",
	"perchunk": true
	},
	"result": "allow"
  },
  {
	"mob": "betteranimalsplus:eel_freshwater",
	"result": "deny"
  }]

**Comment by MechanosG** [comment]
Lmao...  ok so...  I re-added In Control and ran a simple test that should limit all of better animal plus' mobs to 1 per species.
```
	{
		"__comment": "Limit Better Animals Plus to 1 per type (mainly for water creature spam)",
		"mod": "betteranimalsplus",
		"mincount": 1,
		"result": "deny"
	}
```
It technically worked.  But now other mods that have aquatic spawns, are doing the same thing.  Future MC is spamming Salmon everywhere.  So I'm going to go out on a limb, and say this seems to be a bug with In Control and aquatic (or non standard) passive spawns.

**Comment by mxnmnm** [comment]
Did my code for it work? It worked for me so you might have something else that messes with it.

**Comment by MechanosG** [comment]
I didn't test your code specifically.  It looks like it would probably work to limit freshwater eels, but other fish are affected too, including from other mods.

**Comment by mxnmnm** [comment]
Interesting, I didn't experience that at all. Can you post your modlist so I can compare which mods we have in common?

**Comment by MechanosG** [comment]
I don't know how to easily post my entire mod list.  But I'd try [Future MC](https://www.curseforge.com/minecraft/mc-mods/future-mc) and see if the salmon (or other fish) start spamming like the eels.

Edit:  Might not hurt to test Mowzie's Mobs too.  A few people, including me, had problems with it spamming the lantern mobs in forests.  It may be unrelated though.  The author appeared to not be sure why it was happening, eventually blaming it on individual biome mob limits.  I plan on re-testing it now that I know In Control is doing something to some mobs.

**Comment by mxnmnm** [comment]
You could export the modpack from twitch and paste the modlist.html inside of the newly created zip

**Comment by MechanosG** [comment]
Confirmed Mowzie's Mob's lantern spam is also from In Control.

**Comment by mxnmnm** [comment]
Interesting because I've never encountered any issues. Maybe it's some of the specific rules you added in that are causing it? Or are you testing it out with no scripts running?

**Comment by MechanosG** [comment]
I have no rules that generically tell passive mobs that they can spawn.  I have rules to block hostiles from spawning in certain situations, and some rules that prevent too many passive spawns, but nothing that would promote them to spawn more rather than less.

**Comment by mxnmnm** [comment]
I think it would be a good idea to test which rules or if any rules at all cause the issue. Just so I can know what to look for because I did a lot of my rules in other dimensions so I'd like to do some of my own testing and it would be nice to know where to look.

**Comment by MechanosG** [comment]
Alright so, I went ahead and tested without any rules at all.  And curiously the mob replication didn't happen.  But before you assume I just made a bad rule, I'll just show you them.

I have nothing that would tell a mob that it can spawn freely.  That's my entire spawn.json and I don't use any of the other files.  At this point, this tells me that you have to have "some" rules before the problem happens..  but I'm pretty confident it's not a logic issue on my end, and rather..  something breaks after trying to apply a rule.  Like maybe limiting squid spawns causes it glitch or something.

**Comment by mxnmnm** [comment]
This is def an issue with your code because I remember having similar issues with incontrol. I'm not sure how to explain it that well but what you should try doing is making it so that the default action is to deny spawns of the target mob/mobs but spawns should be allowed under certain circumstances. What you're doing is making it allowed by default and having it be denied under certain circumstances. 

tl;dr- just deny the mob spawns by default then allow it under certain circumstances

**Comment by MechanosG** [comment]
You're wrong.  In Control is based on the vanilla spawn system, not it's own.  Aka, it's not going to just spam infinitely if you don't deny everything.  You deny requests when you don't want the vanilla spawn system to spawn how it would normally.  Anytime you send it an "allow", it will bypass all vanilla checks and only filter based on whatever your rule specifically states.  That's why I tell it to use "Default" when I make an exception for the squid to spawn, so it'll use regular vanilla spawning.  Aka, if Vanilla says squid aren't allowed to spawn there, it'll not spawn there.

Try again.

**Comment by itsmeow** [comment]
IF YOU COMMENT HERE ONE MORE TIME I AM CLOSING THIS ISSUE. MY INBOX IS DEAD. DISCORD. USE DISCORD.

**Comment by MechanosG** [comment]
Might as well close it.  In Control caused the problem for me, but he's trying to blame me (yet can't provide an example) based on poor understanding of how In Control works.  I'm out.  Sorry you got spammed @itsmeow .

**Comment by itsmeow** [comment]
Yeah I really think this is an issue with InControl and water spawns.
Follow these two issues: https://github.com/McJtyMods/InControl/issues/175 https://github.com/McJtyMods/InControl/issues/191

**Comment by PiggiesGoSqueal** [comment]
> 
> 
> Confirmed Mowzie's Mob's lantern spam is also from In Control.



> 
> 
> Yeah I really think this is an issue with InControl and water spawns.
> Follow these two issues: [McJtyMods/InControl#175](https://github.com/McJtyMods/InControl/issues/175) [McJtyMods/InControl#191](https://github.com/McJtyMods/InControl/issues/191)


Just to clarify, does that mean these mods (& any others that use water spawns) won't work alongside InControl? Or just that you can't use custom spawn settings for these mods using InControl?

**Comment by itsmeow** [comment]
I'm not sure. Not familiar with InControl's internals but if they replace spawns regardless of any configuration, they will not work. If you disable InControl on a water spawn it may or may not fix it. No idea.
